CLOUD: Add Support for Emscripten

chkuendig opened the pull request:
This is adding the option to bring your own games to the web version of ScummVM: https://www.youtube.com/watch?v=kYechxc5tvE

I was aiming for a as smooth and "web-native" integration as possible, while still using the existing infrastructure relying on cloud.scummvm.org, with the end-result looking like any other Oauth-flow: A new window opens, the user authenticates the app, and once done the window closes and the process continues in the original page. 

A few things:
- I'll send a PR for the cloud component once we are closer to actually deploying it. The current change is minimal: https://github.com/chkuendig/scummvm-cloud/commit/bdba427927e7eb172d94f1af627ca845ff1e90c0.
- I'm still looking into how I could add range-request support. Many games have very large asset packs (100M+), and these need to get fully downloaded before launching the game or sometimes even before adding it. But I expect this to be a pretty large change across all storage backends, so I wanted to first get the basics approved before continuing
- ~I tried to stick as closely to the current curl-based implementation and refrained from adding another abstraction layer, as I assumed Emscripten most likely being the only platform with network but not libcurl.~  _**see bellow, I refactored it to isolate the platform-specific code.**_
  - ~One result of this is that I actually just re-used the existing curl code for storing the header structure. If needed I could re-implement this, so let me know if there's an issue with the current setup~. **_fixed_**
- ~CI is broken, seems like my syntax for checking wether either `USE_LIBCURL` or `EMSCRIPTEN` is defined isn't working on all platforms.~ **_fixed_**

Generally I expect this to need quite a bit of additional work, but I was hoping to get feedback early.

chkuendig commented on backends/cloud/box/boxstorage.cpp:
```diff
@@ -33,6 +32,11 @@
 #include "common/config-manager.h"
 #include "common/debug.h"
 #include "common/formats/json.h"
+#ifdef USE_LIBCURL
+#include <curl/curl.h>
+#elif defined(EMSCRIPTEN)
+#include "backends/networking/curl/emscripten/slist.h"
+#endif
```
Yes, I'll try to add a common header that avoids having these switches everywhere.

chkuendig replied:
I removed all references to curl_slist outside of the curl-specific implementation.

lephilousophe commented on engines/testbed/cloud.cpp:
```diff
@@ -441,9 +441,12 @@ TestExitStatus CloudTests::testDownloading() {
-	Common::FSDirectory gameRoot(path);
-	Common::FSNode node = gameRoot.getFSNode().getChild("downloaded_file.txt");
+	Common::FSNode testDirectory(ConfMan.getPath("path"));
+	if (!testDirectory.isWritable()) {
```
In fact, I believe this should never try to write in the game folder.

chkuendig commented:
@lephilousophe I implemented most of what we discussed. I still kept NetworkReadStream to a single class, as the #ifdefs are quite limited (basically a few class variables and some callbacks are different). 

I also renamed all the shared parts to better separate what's curl-specific and what's common - which is much clearer now because the request headers aren't stored in `curl_slist` anymore.

chkuendig commented on engines/testbed/config-params.h:
```diff
@@ -53,7 +53,7 @@ class ConfigParams : public Common::Singleton<ConfigParams> {
 	bool _isInteractive;
 	bool _isGameDataFound;
-#ifdef USE_LIBCURL
+#if defined(USE_LIBCURL) || defined(EMSCRIPTEN) 
```
thanks, merged/squashed it.

chkuendig commented on backends/fs/emscripten/emscripten-fs-factory.cpp:
```diff
@@ -80,6 +84,10 @@ AbstractFSNode *EmscriptenFilesystemFactory::makeFileNodePath(const Common::Stri
 		return new HTTPFilesystemNode(*(_httpNodes->getVal(path)));
+#ifdef USE_CLOUD
+    } else if (path.hasPrefix(CLOUD_FS_PATH) && CloudMan.isStorageEnabled()) {
+        return new CloudFilesystemNode(path);
```
thanks, merged/squashed it.

ccawley2011 reviewed (commented):
There are some `USE_LIBCURL` checks in the backends/networking/sdl_net that can also be removed or replaced with `USE_CLOUD`.

chkuendig commented on backends/networking/http/connectionmanager.cpp:
```diff
@@ -109,11 +123,22 @@ Common::String ConnectionManager::urlEncode(const Common::String &s) const {
-		return result;
 	}
-	return "";
+#elif defined(EMSCRIPTEN)
```
yes, I decided to refactor ConnectionManager and NetworkReadStream today. @lephilousophe already recommended doing so.

ccawley2011 replied:
You could probably remove this specific use of platform-specific APIs by using `Common::percentEncodeString()` here instead.

chkuendig replied:
I just tested it, and unfortunately this function doesn't fully support Unicode. 

I have a folder called `Ultima IV™.app` (the GOG macOS release) on my OneDrive. Correctly url encoded, this would be `Ultima%20IV%E2%84%A2.app`, but Common::percentEncodeString turns this into `Ultima%20IV%FFFFFFE2%FFFFFF84%FFFFFFA2.app` which won't work.

Tkachov commented on backends/networking/http/emscripten/networkreadstream-emscripten.cpp:
```diff
@@ -0,0 +1,307 @@
+	unsigned short responseCode = 0;
+	if (_emscripten_fetch)
+		responseCode = _emscripten_fetch->status;
+	debug(5, "NetworkReadStreamEmscripten::httpResponseCode %hu %hu ", _emscripten_fetch->status, responseCode);
```
checking `_emscripten_fetch` above, but accessing here regardless

Tkachov commented on devtools/create_project/create_project.cpp:
```diff
@@ -1297,6 +1298,16 @@ static void fixupFeatures(ProjectType projectType, BuildSetup &setup) {
+	// With libcurl we get an HTTP client
+	if (!getFeatureBuildState("libcurl", setup.features)) {
+		setFeatureBuildState("http", setup.features, false);
+	}
```
should this be
```
if (getFeatureBuildState("libcurl", setup.features)) {
	setFeatureBuildState("http", setup.features, true);
}
```
according to comment?

should there be another `if` that checks for emscripten and fixes value of http too?

lephilousophe replied:
http feature is on by default so not having libcurl makes it disabled.
I will change the comment.

For emscripten, no need, as the platform doesn't use create_project I suppose.

Tkachov reviewed (commented):
LGTM

left a couple of minor comments/questions

lephilousophe commented on backends/networking/http/networkreadstream.h:
```diff
@@ -0,0 +1,189 @@
+		: _backingStream(DisposeAfterUse::YES), _keepAlive(keepAlive), _keepAliveIdle(keepAliveIdle), _keepAliveInterval(keepAliveInterval),
+		  _eos(false), _requestComplete(false), _sendingContentsBuffer(nullptr), _sendingContentsSize(0), _sendingContentsPos(0),
+		  _progressDownloaded(0), _progressTotal(0) {
+	}
```
Indeed, thank you. It's applied.
